### PR TITLE
BUG, MAINT: check that histogram range parameters are finite.

### PR DESCRIPTION
**Pull request by lzkelley**
This assures that the values in the `range` parameter are finite, avoiding meaningless results.  See [Scipy Issue #5221](https://github.com/matplotlib/matplotlib/issues/5221), and [[Numpy-discussion]histogram gives meaningless results with non-finite range](https://mail.scipy.org/pipermail/numpy-discussion/2015-November/074161.html).

### CODE REVIEW

**Comment by lzkelley** [comment]
Thanks for the suggestions!  I've fixed the PEP8 issue, and added tests.

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -338,6 +338,10 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
         if (mn > mx):
             raise AttributeError(
                 'max must be larger than min in range parameter.')
+        if not np.all(np.isfinite([mn, mx])):
+            raise AttributeError(
```
Comment (jaimefrio) [inline numpy/lib/function_base.py]: I suppose you have tried to replicate what was already there, but this should be a `ValueError`, since the problem is not that ["an attribute reference or assignment fails,"](https://docs.python.org/3.5/library/exceptions.html#AttributeError) but ["an inappropriate value."](https://docs.python.org/3.5/library/exceptions.html#ValueError)

You don't have to change it in this PR, but I think the `AttributeError` on line 339 should be changed to a `ValueError` as well.

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -551,6 +555,9 @@ def histogramdd(sample, bins=10, range=None, normed=False, weights=None):
             smin = atleast_1d(array(sample.min(0), float))
             smax = atleast_1d(array(sample.max(0), float))
     else:
+        if not np.all(np.isfinite(range)):
+            raise AttributeError(
```
Comment (jaimefrio) [inline numpy/lib/function_base.py]: Same as above, should be `ValueError`.

File: numpy/lib/function_base.py (hunk 1 above)
Comment (lzkelley) [inline numpy/lib/function_base.py]: That makes sense.  I'm happy to add that change here.  Should that be its own PR, instead?

Comment (jaimefrio) [inline numpy/lib/function_base.py]: Not really, can be put in this one if you want too. But you may have to go fixing failing tests after the change, if it turns into too much work, we can simply open an issue and come back to it later.

Comment (lzkelley) [inline numpy/lib/function_base.py]: Went ahead and changed these to `ValueError`s.

File: numpy/lib/tests/test_function_base.py [hunk 3]
```diff
@@ -1267,6 +1267,14 @@ def test_empty(self):
         assert_array_equal(a, np.array([0]))
         assert_array_equal(b, np.array([0, 1]))
 
+    def test_finite_range(self):
+        # Normal ranges should be fine
+        vals = np.linspace(0.0, 1.0, num=100)
+        histogram(vals, range=[0.25,0.75])
+        # non-finite values should raise AttributeError
```
Comment (jaimefrio) [inline numpy/lib/tests/test_function_base.py]: This is why (some) people sneer at comments: it wasn't updated and now contradicts the code.  I'll leave it up tp you whether to remove it entirely (the code is pretty much self explanatory, I think) or to update it to `ValueError`.

File: numpy/lib/tests/test_function_base.py [hunk 4]
```diff
@@ -1489,6 +1497,18 @@ def test_rightmost_binedge(self):
         assert_(hist[0] == 0.0)
         assert_(hist[1] == 0.0)
 
+    def test_finite_range(self):
+        vals = np.random.random((100,3))
+        # Normal ranges should be fine
+        histogramdd(vals, range=[[0.0,1.0],[0.25,0.75],[0.25,0.5]])
+        # non-finite values should raise AttributeError
```
Comment (jaimefrio) [inline numpy/lib/tests/test_function_base.py]: Same comment mishap here...

File: numpy/lib/tests/test_function_base.py [hunk 5]
```diff
@@ -1267,6 +1267,14 @@ def test_empty(self):
         assert_array_equal(a, np.array([0]))
         assert_array_equal(b, np.array([0, 1]))
 
+    def test_finite_range(self):
+        # Normal ranges should be fine
+        vals = np.linspace(0.0, 1.0, num=100)
+        histogram(vals, range=[0.25,0.75])
+        # non-finite values should raise AttributeError
+        assert_raises(ValueError, histogram, vals, range=[np.nan,0.75])
+        assert_raises(ValueError, histogram, vals, range=[np.inf,0.75])
```
Comment (jaimefrio) [inline numpy/lib/tests/test_function_base.py]: Might be a good idea to have non-finite values in the second entry of the `range` in some of the tests, in case a future change misses one of the entries.

**Comment by jaimefrio** [comment]
Thanks for fixing the `AttirbuteError` thing for the general case.  i have commented a couple of loose ends on the tests, but it is about ready to go.  If after fixing those you squash all the commits into a single one, I think this will be ready to go.

**Comment by lzkelley** [comment]
I used `rebase -i` to squash into a single commit (I hope that's the right approach), after fixing those documentation issues.  Should be good!

**Comment by jaimefrio** [comment]
Thanks Luke, in it goes.
